aot/c_codegen.py

```diff
@@ -0,0 +1,213 @@
+from pathlib import Path
+from dataclasses import dataclass
+import binascii
+from typing import Sequence, Tuple
+
+
+from abstract_values import AbstractPtr, AbstractValue
+
+@dataclass
+class LibCudaConf:
+    include_dir: str
+    link_dir: str
+    
+
+@dataclass
+class CSources:
+    sources: Sequence[str]
+    headers: Sequence[str]
+    lib_name: str
+    
+    def iter_header_and_sources(self):
+        for hdr, src in zip(self.headers, self.sources):
+            yield hdr, src
+
+@dataclass
+class CInputSigniture:
+    """Build kernel specific function signiture"""
+
+    # TODO: make sure the mapping is correct
+    _TRITON_TO_C_TYPE = {
+        "I": "int32_t",
+        "f": "float",
+        "B": "bool",
+        "f8": "float",
+        "f16": "float",
+        "bf16": "float",
+        "f32": "float",
+        "f64": "double",
+        "i1": "bool",
+        "i8": "int8_t",
+        "i16": "int16_t",
+        "i32": "int32_t",
+        "i64": "int64_t",
+    }
+
+    input_types: Sequence[AbstractValue]
+    arg_names: Sequence[str]
+
+    def __len__(self):
+        return len(self.input_types)
+
+    def signitue_c_code(self) -> str:
+        sig = ""
+        for arg, ty in zip(self.arg_names, self.input_types):
+            if isinstance(ty, AbstractPtr):
+                dtype = "CUdeviceptr"
+            else:
+                dtype = CInputSigniture._TRITON_TO_C_TYPE[ty.tt_dtype]
+            sig = f"{sig}{dtype} {arg}, "
+        if len(sig):
+            sig = sig[:-2]
+        return sig
+
+
+def py_str_to_uchar_array(txt: str) -> Tuple[str, int]:  # (c_code, array len)
+    """Hexdump as string into a C array"""
+    hex_ = str(binascii.hexlify(bytes(txt, "utf-8")))[2:-1]
+    it = iter(hex_)
+    data = ", ".join([f"0x{x}{next(it)}" for x in it])
+    return data, len(hex_)
+
+
+# CUdeviceptr x, CUdeviceptr y, CUdeviceptr res, size_t size
+def gen_kernel_code(name: str, ptx_src: str, data_sign: CInputSigniture) -> str:
+    ptx_as_hex, ptx_size = py_str_to_uchar_array(ptx_src)
+
+    ptx_def = f"unsigned char {name}_ptx[{ptx_size}]"
```

Comment:
I think here we probably want to save the bytes of the raw binary CUDA object. These can also be loaded with cuModuleLoad but this bypasses the Driver's JIT compilation from PTX to SASS. Right now I don't remember if there's a way to query that in Python, but I could add it.